app/src/test/java/de/westnordost/streetcomplete/util/SpatialCacheTest.kt

```diff
@@ -0,0 +1,92 @@
+package de.westnordost.streetcomplete.util
+
+import de.westnordost.streetcomplete.data.download.tiles.enclosingTilePos
+import de.westnordost.streetcomplete.data.osm.mapdata.LatLon
+import de.westnordost.streetcomplete.data.osm.mapdata.Node
+import de.westnordost.streetcomplete.testutils.node
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertFalse
+import org.junit.Assert.assertNull
+import org.junit.Assert.assertTrue
+import org.junit.Test
+
+internal class SpatialCacheTest {
```

Comment:
I have a questrion regarding `get returns all the data in the bbox`:
I try adding nodes to the same tile, but at at different positions. This works, but
```kotlin
        val fetchTile = LatLon(1.0, 1.0).enclosingTilePos(16)
        val fetchTileBBox = fetchTile.asBoundingBox(16)
        val nodesInsideFetchBBox = listOf(
            node(1, LatLon(fetchTileBBox.min.latitude + 0.0001, fetchTileBBox.min.longitude + 0.0001)),
            node(2, LatLon(fetchTileBBox.min.latitude + 0.0002, fetchTileBBox.min.longitude + 0.0002))
        )
        val fetchBBox = nodesInsideFetchBBox.map { it.position }.enclosingBoundingBox()
        assertTrue(nodesInsideFetchBBox[0].position in fetchBBox)
        assertTrue(nodesInsideFetchBBox[1].position in fetchBBox)
```
fails on the second `assertTrue`.
Why is that happening? Is it expected that `enclosingBoundingBox` creates a BBox that doesn't contain all positions in the list, or is it a bug?
I'd rather assume the second, as `assertTrue(nodesInsideFetchBBox[1].position.longitude > fetchBBox.max.longitude)` passes.

Reply:
I have trouble understanding what exactly the code snippet above tests. 

But anyway `Iterable<LatLon>.enclosingBoundingBox()` isn't very thoroughly tested. There is only one test to that effect and that is
```kotlin
    @Test fun `enclosingbbox for line`() {
        val positions = listOf(p(0.0, -4.0), p(3.0, 12.0), p(16.0, 1.0), p(-6.0, 0.0))
        val bbox = positions.enclosingBoundingBox()
        assertEquals(-4.0, bbox.min.latitude, 0.0)
        assertEquals(12.0, bbox.max.latitude, 0.0)
        assertEquals(16.0, bbox.max.longitude, 0.0)
        assertEquals(-6.0, bbox.min.longitude, 0.0)
    }
```

Maybe there is some situation it did not catch?

Reply:
I suspect a floating point imprecision that is causing this. You know, the famous
```
>>> 0.1+0.2
0.30000000000000004
```

Reply:
> I have trouble understanding what exactly the code snippet above tests.

Well, it's only the beginning of some test, where I then added some ´assertTrue` to see where things go wrong.

Problem can be reproduced with
```kotlin
    @Test fun `enclosingbbox for points`() {
        val pointInsideBBox = LatLon(0.0002, 0.0002)
        val pointBBox = listOf(pointInsideBBox).enclosingBoundingBox()
        assertTrue(pointInsideBBox in pointBBox)
    }
```
Though for now I have no idea how to fix it....

Reply:
I'll take care of this then, it's my bug after all. Thank you!

Reply:
weird, 
```
    @Test fun `enclosingbbox for points`() {
        val p = p(0.0002, 0.0002)
        val bbox = listOf(p).enclosingBoundingBox()
        assertTrue(bbox.contains(p))
    }
```
does not fail on master

Reply:
This works, but fails when changing to `val p = LatLon(0.0002, 0.0002)`
Looks like it only works when `normalizeLongitude` is used.

Reply:
I removed a point where floating point imprecision is introduced, but the test ran successfully both before and after. So I don't know why it fails for you... do the test I added fail for you too?

Reply:
> This works, but fails when changing to val p = LatLon(0.0002, 0.0002)
> Looks like it only works when normalizeLongitude is used.

Can't reproduce that on master

Reply:
Anyway, I have a fix ready that should not introduce any floating point imprecisions, but it would be nice if that would be backed up by a test that the current implementation would not pass.

```diff
--- a/app/src/main/java/de/westnordost/streetcomplete/util/math/SphericalEarthMath.kt
+++ b/app/src/main/java/de/westnordost/streetcomplete/util/math/SphericalEarthMath.kt
@@ -218,22 +218,33 @@ fun Iterable<LatLon>.enclosingBoundingBox(): BoundingBox {
     var minLonOffset = 0.0
     var maxLatOffset = 0.0
     var maxLonOffset = 0.0
+    var minLat = origin.latitude
+    var minLon = origin.longitude
+    var maxLat = origin.latitude
+    var maxLon = origin.longitude
     while (it.hasNext()) {
         val pos = it.next()
         // calculate with offsets here to properly handle 180th meridian
         val latOffset = pos.latitude - origin.latitude
         val lonOffset = normalizeLongitude(pos.longitude - origin.longitude)
-        if (latOffset < minLatOffset) minLatOffset = latOffset
-        if (lonOffset < minLonOffset) minLonOffset = lonOffset
-        if (latOffset > maxLatOffset) maxLatOffset = latOffset
-        if (lonOffset > maxLonOffset) maxLonOffset = lonOffset
+        if (latOffset < minLatOffset) {
+            minLatOffset = latOffset
+            minLat = pos.latitude
+        }
+        if (lonOffset < minLonOffset) {
+            minLonOffset = lonOffset
+            minLon = pos.longitude
+        }
+        if (latOffset > maxLatOffset) {
+            maxLatOffset = latOffset
+            maxLat = pos.latitude
+        }
+        if (lonOffset > maxLonOffset) {
+            maxLonOffset = lonOffset
+            maxLon = pos.longitude
+        }
     }
-    return BoundingBox(
-        origin.latitude + minLatOffset,
-        normalizeLongitude(origin.longitude + minLonOffset),
-        origin.latitude + maxLatOffset,
-        normalizeLongitude(origin.longitude + maxLonOffset)
-    )
+    return BoundingBox(minLat, minLon, maxLat, maxLon)
 }

```

Reply:
Looks good, everything works on current master, thanks!

Reply:
> it would be nice if that would be backed up by a test that the current implementation would not pass. I do not seem to be able to reproduce any.

For me this failed:
```kotlin
    @Test fun `enclosingbbox for points`() {
        val p = LatLon(0.0002, 0.0002)
        val bbox = listOf(p).enclosingBoundingBox()
        assertTrue(bbox.contains(p))
    }
```
While both of these worked:
```kotlin
    @Test fun `enclosingbbox for points`() {
        val p = p(0.0002, 0.0002)
        val bbox = listOf(p).enclosingBoundingBox()
        assertTrue(bbox.contains(p))
    }
    @Test fun `enclosingbbox for points2`() {
        val p = LatLon(0.0002, normalizeLongitude(0.0002))
        val bbox = listOf(p).enclosingBoundingBox()
        assertTrue(bbox.contains(p))
    }
```

Reply:
For me, it does not fail (on master). This is super-weird. It should not even matter if you are on Windows, Linux or Mac, after all, it all runs in the JVM. Maybe the Java version matters? I couldn't think of anything else that mattered.

I use Java version 11.0.12 for the tests. (I don't know where to look that up in the project settings in the IDE, but you can access it via `System.getProperty("java.version")` while the test is running (set breakpoint somewhere in the test, press Alt+F8 and write what you want to execute).

Reply:
I think you can also look it up here:

![image](https://user-images.githubusercontent.com/4661658/180524037-50ef80e8-d511-4780-b455-7ffab52e3f03.png)

Reply:
Also using 11.0.12, according to both settings and `System.getProperty("java.version")`
```kotlin
    @Test fun `enclosingbbox for points`() {
        val p = LatLon(0.0002, 0.0002)
        val bbox = listOf(p).enclosingBoundingBox()
        assertTrue(bbox.contains(p))
    }
```
Passes after https://github.com/streetcomplete/StreetComplete/commit/4257991aaba9b966b41ba582b20d729b9e87fe41, but fails before